pkg/coscheduling/coscheduling.go

```diff
@@ -95,294 +83,364 @@ func New(obj runtime.Object, handle framework.FrameworkHandle) (framework.Plugin
+			cs.pgMgr.PodGroupStatusCache().Set(util.GetPodGroupFullName(pgs.PodGroup), pgs)
 		}
 	})
 }
 
-// GetPodGroupLabels checks if the pod belongs to a PodGroup. If so, it will return the
-// podGroupName, minAvailable of the PodGroup. If not, it will return "" and 0.
-func GetPodGroupLabels(pod *v1.Pod) (string, int, error) {
-	podGroupName, exist := pod.Labels[PodGroupName]
-	if !exist || len(podGroupName) == 0 {
-		return "", 0, nil
+// StartBatchSchedule receives the group name of pod to start scheduling
+func (cs *Coscheduling) StartBatchSchedule(fullName string) {
+	pgs := cs.pgMgr.PodGroupStatusCache().Get(fullName)
+	if pgs.PodGroup.Status.Phase != v1alpha1.PodGroupPreScheduling && pgs.PodGroup.Status.Phase != v1alpha1.
+		PodGroupScheduling {
+		return
+	}
+
+	// record time to avoid abnormal exit when bind
+	if pgs.PodGroup.Status.Scheduled >= pgs.PodGroup.Spec.MinMember {
+		pg := pgs.PodGroup
+		pgCopy := pgs.PodGroup.DeepCopy()
+		pgCopy.Status.ScheduleStartTime = metav1.Now()
+
+		patch, err := util.CreateMergePatch(pg, pgCopy)
+		if err != nil {
+			return
+		}
+
+		if err := cs.pgMgr.PatchPodGroup(pg.Name, pg.Namespace, patch); err != nil {
+			return
+		}
 	}
-	minAvailable, exist := pod.Labels[PodGroupMinAvailable]
-	if !exist || len(minAvailable) == 0 {
-		return "", 0, nil
+
+	// Start coscheduling
+	klog.V(4).Infof("Start scheduling %v", fullName)
+	pendingPods := cs.pgMgr.GetPodNodePairs(fullName)
+	if pendingPods == nil {
+		klog.V(4).Infof("Can not found pending pods for %v", fullName)
+		return
 	}
-	minNum, err := strconv.Atoi(minAvailable)
-	if err != nil {
-		klog.Errorf("PodGroup %v/%v : PodGroupMinAvailable %v is invalid", pod.Namespace, pod.Name, minAvailable)
-		return "", 0, err
+	pendingPodNameIDs := cs.pgMgr.GetAssumedPods(fullName)
+	if pendingPodNameIDs == nil {
+		klog.V(4).Infof("Can not found pending IDs for %v", fullName)
+		return
 	}
-	if minNum < 1 {
-		klog.Errorf("PodGroup %v/%v : PodGroupMinAvailable %v is less than 1", pod.Namespace, pod.Name, minAvailable)
-		return "", 0, err
+	pendingPodsMap := pendingPods.Items()
+	if int32(len(pendingPodsMap)) < pgs.PodGroup.Spec.MinMember-pgs.PodGroup.Status.Scheduled {
+		return
 	}
-	return podGroupName, minNum, nil
-}
 
-func (cs *Coscheduling) calculateTotalPods(podGroupName, namespace string) int {
-	// TODO get the total pods from the scheduler cache and queue instead of the hack manner.
-	selector := labels.Set{PodGroupName: podGroupName}.AsSelector()
-	pods, err := cs.podLister.Pods(namespace).List(selector)
-	if err != nil {
-		klog.Error(err)
-		return 0
+	// firstPodUid just record the first pod in the podsMap. When we delete any one of the pending pods,
+	// it would trigger evict of all cache.
+	firstPodUid := ""
+	for uid, pair := range pendingPodsMap {
+		waitingPod := cs.frameworkHandler.GetWaitingPod(types.UID(uid))
+		if waitingPod == nil {
+			// to avoid sig send, but scheduler cache have not been flushed
```

Comment:
I don't understand this comment.

Reply:
have removed, here we are just checking if waitingPod is nil